File: src/Examples/pres/ex_cs.m
%% ML-JL SPGL Comparison
clear

mypath = '/home/Software/slim/';
addpath(genpath([mypath 'SLIM-release-apps/tools/solvers/GenSPGL1/']));

% Define Size of problem
%n  = 1024;
%k  = 50; 

% Sparse Vector x0
%x0 = zeros(n,1);
%x0(randperm(n, 10)) = randn(10,1);
%r_inds = randperm(n, k);

% Load sparse vector and restriction mask
load('x0.mat')
n = length(x0);

% Sparsifying Transform
S  = opDCT(n);

% Restriction Operator
R  = opRestriction(n, r_inds);

% Modelling Operator
A = R*S

% Create Data
b  = A*x0;

% Solve
tic
[x,r,g,info] = spgl1(A, b);
t_ml = toc;
SNR = -20*log10(norm(x0-x)/norm(x0));

%Compare
load('x_jl.mat')
figure(1)
subplot 121
    plot(1:length(x), x, 'k-', 'LineWidth', 2); hold on
    plot(1:length(x_jl), x_jl,'r--', 'LineWidth', 2)
    legend('MATLAB', 'JULIA')
    title(['Solution, Julia Speedup:' num2str(round(t_ml/t_jl, 2))])
subplot 122
    plot(x-x_jl)
    title('Residual')

File: src/Examples/pres/ex_cs.jl
## JL-ML SPGL Comparison
export ex_cs

"""
An example of using implicit JOLI operators to solve a system using GenSPGL

Use: x, r, g, info, SNR = ex_cs()
"""
function ex_cs()
    #n = 1024
    #k = 50

    # Create a sparse vector
    #inds = randperm(n)
    #x0 = zeros(n)
    #x0[inds[1:10]] = randn(10)

    # Load sparse vector x and restriction mask
    fid = matopen(Pkg.dir("GenSPGL")*"/src/Examples/pres/x0.mat")
    x0 = vec(read(fid, "x0"))
    r_inds = Int.(vec(read(fid, "r_inds")))
    n = length(vec(x0))
    close(fid)

    # Sparsity Promoting Transform
    S = joDCT(n)
    
    # Restriction Operator
    R = joRestriction(n, r_inds)
    
    # Modelling op
    A = R*S

    # Create data
    b = A*x0

    # Set BLAS Threads
    BLAS.set_num_threads(8)

    # Solve
    opts = spgOptions(verbosity = 1)
    x, r, g, info = spgl1(A, b, options = opts) 

    # Solve again silently for timing
    opts = spgOptions(verbosity = 0)
    t_jl = @elapsed spgl1(A, b, options = opts)

    # Calc SNR of recovery
    SNR = -20*log10(norm(x0-x)/norm(x0));

    # Write results
    fw = matopen("x_jl.mat", "w")
    write(fw, "x_jl", x)
    write(fw, "t_jl", t_jl)
    close(fw)
   
    return x, r, g, info, SNR
end
